ruff_lsp/server.py

```diff
@@ -679,99 +1059,183 @@ async def resolve_code_action(params: CodeAction) -> CodeAction:
+        edits = _fixed_source_to_edits(
+            original_source=document.source, fixed_source=result.stdout.decode("utf-8")
+        )
+        return WorkspaceEdit(
+            document_changes=[
+                _create_text_document_edit(document.uri, document.version, edits)
+            ]
+        )
+    elif document.kind is DocumentKind.Notebook:
+        notebook_document = LSP_SERVER.workspace.get_notebook_document(
+            notebook_uri=document.uri
+        )
+        if notebook_document is None:
+            log_warning(f"No notebook document found for {document.uri!r}")
+            return None
+
+        output_notebook_cells = cast(
+            Notebook, json.loads(result.stdout.decode("utf-8"))
+        )["cells"]
+        if len(output_notebook_cells) != len(notebook_document.cells):
+            log_warning(
+                f"Number of cells in the output notebook doesn't match the number of "
+                f"cells in the input notebook. Input: {len(notebook_document.cells)}, "
+                f"Output: {len(output_notebook_cells)}"
+            )
+            return None
+
+        cell_document_changes: list[TextDocumentEdit] = []
+        for cell_idx, cell in enumerate(notebook_document.cells):
+            if cell.kind is not NotebookCellKind.Code:
+                continue
+            cell_document = LSP_SERVER.workspace.get_text_document(cell.document)
+            edits = _fixed_source_to_edits(
+                original_source=cell_document.source,
+                fixed_source=output_notebook_cells[cell_idx]["source"],
+            )
+            cell_document_changes.append(
+                _create_text_document_edit(
+                    cell_document.uri,
+                    cell_document.version,
+                    edits,
+                )
+            )
+
+        return WorkspaceEdit(document_changes=list(cell_document_changes))
+    elif document.kind is DocumentKind.Cell:
+        text_edits = _result_single_cell_notebook_to_edits(document, result)
+        if text_edits is None:
+            return None
+        return WorkspaceEdit(
+            document_changes=[
+                _create_text_document_edit(document.uri, document.version, text_edits)
+            ]
+        )
+    else:
+        assert_never(document.kind)
 
-    new_source = _match_line_endings(document, result.stdout.decode("utf-8"))
 
-    # Skip last line ending in a notebook cell.
-    if document.uri.startswith("vscode-notebook-cell"):
-        if new_source.endswith("\r\n"):
-            new_source = new_source[:-2]
-        elif new_source.endswith("\n"):
-            new_source = new_source[:-1]
+def _result_single_cell_notebook_to_edits(
+    document: Document, result: RunResult
+) -> list[TextEdit] | None:
+    """Converts a run result to a list of TextEdits.
 
-    if new_source == document.source:
+    The result is expected to be a single cell Notebook Document.
+    """
```

Comment:
This function exists because for formatting provider, we need to return a list of `TextEdit` while for the command we need to apply a `WorkspaceEdit`.